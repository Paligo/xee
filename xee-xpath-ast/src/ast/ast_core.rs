use ibig::IBig;
use ordered_float::OrderedFloat;
use rust_decimal::prelude::*;
use xot::Xot;

pub use crate::operator::BinaryOperator;
use crate::{
    namespaces::Namespaces,
    span::{Spanned, WithSpan},
};

pub type ExprSingleS = Spanned<ExprSingle>;
pub type PrimaryExprS = Spanned<PrimaryExpr>;
pub type StepExprS = Spanned<StepExpr>;
pub type ExprS = Spanned<Expr>;
pub type NameS = Spanned<Name>;

impl WithSpan for ExprSingle {}
impl WithSpan for PrimaryExpr {}
impl WithSpan for StepExpr {}
impl WithSpan for Expr {}
impl WithSpan for Name {}

#[derive(Debug, Clone, PartialEq, Eq)]
#[cfg_attr(test, derive(serde::Serialize))]
pub struct Expr(pub Vec<ExprSingleS>);

#[derive(Debug, Clone, PartialEq, Eq)]
#[cfg_attr(test, derive(serde::Serialize))]
pub struct XPath(pub ExprS);

#[derive(Debug, Clone, PartialEq, Eq)]
#[cfg_attr(test, derive(serde::Serialize))]
pub enum ExprSingle {
    // a path expression
    Path(PathExpr),
    // something applied to a path expression
    Apply(ApplyExpr),
    // combine two path expressions
    Let(LetExpr),
    If(IfExpr),
    Binary(BinaryExpr),
    For(ForExpr),
    Quantified(QuantifiedExpr),
}

#[derive(Debug, Clone, PartialEq, Eq)]
#[cfg_attr(test, derive(serde::Serialize))]
pub struct ForExpr {
    pub var_name: NameS,
    pub var_expr: Box<ExprSingleS>,
    pub return_expr: Box<ExprSingleS>,
}

#[derive(Debug, Clone, PartialEq, Eq)]
#[cfg_attr(test, derive(serde::Serialize))]
pub struct QuantifiedExpr {
    pub quantifier: Quantifier,
    pub var_name: NameS,
    pub var_expr: Box<ExprSingleS>,
    pub satisfies_expr: Box<ExprSingleS>,
}

#[derive(Debug, Clone, PartialEq, Eq, Hash)]
#[cfg_attr(test, derive(serde::Serialize))]
pub struct Name {
    name: String,
    prefix: Option<String>,
    namespace: Option<String>,
}

impl Name {
    pub fn new(name: String, namespace: Option<String>) -> Self {
        Name {
            name,
            namespace,
            prefix: None,
        }
    }

    pub fn prefixed(prefix: &str, name: &str, namespaces: &Namespaces) -> Option<Self> {
        let namespace = namespaces.by_prefix(prefix)?;
        Some(Name {
            name: name.to_string(),
            namespace: Some(namespace.to_string()),
            prefix: Some(prefix.to_string()),
        })
    }

    pub fn unprefixed(name: &str) -> Self {
        Name {
            name: name.to_string(),
            namespace: None,
            prefix: None,
        }
    }

    pub fn uri_qualified(uri: &str, name: &str) -> Self {
        Name {
            name: name.to_string(),
            namespace: Some(uri.to_string()),
            prefix: None,
        }
    }

    pub fn namespace(&self) -> Option<&str> {
        self.namespace.as_deref()
    }

    pub fn local_name(&self) -> &str {
        &self.name
    }

    pub fn to_name_id(&self, xot: &Xot) -> Option<xot::NameId> {
        if let Some(namespace) = &self.namespace {
            let namespace_id = xot.namespace(namespace);
            if let Some(namespace_id) = namespace_id {
                xot.name_ns(&self.name, namespace_id)
            } else {
                None
            }
        } else {
            xot.name(&self.name)
        }
    }

    pub fn with_suffix(&self) -> Name {
        let mut name = self.name.clone();
        name.push('*');
        Name {
            name,
            namespace: self.namespace.clone(),
            prefix: self.prefix.clone(),
        }
    }
}

#[derive(Debug, Clone, PartialEq, Eq)]
#[cfg_attr(test, derive(serde::Serialize))]
pub struct LetExpr {
    pub var_name: NameS,
    pub var_expr: Box<ExprSingleS>,
    pub return_expr: Box<ExprSingleS>,
}

#[derive(Debug, Clone, PartialEq, Eq)]
#[cfg_attr(test, derive(serde::Serialize))]
pub struct IfExpr {
    pub condition: ExprS,
    pub then: Box<ExprSingleS>,
    pub else_: Box<ExprSingleS>,
}

#[derive(Debug, Clone, PartialEq, Eq)]
#[cfg_attr(test, derive(serde::Serialize))]
pub enum Quantifier {
    Some,
    Every,
}

#[derive(Debug, Clone, PartialEq, Eq)]
#[cfg_attr(test, derive(serde::Serialize))]
pub enum PrimaryExpr {
    Literal(Literal),
    VarRef(Name),
    Expr(ExprS),
    ContextItem,
    FunctionCall(FunctionCall),
    NamedFunctionRef(NamedFunctionRef),
    InlineFunction(InlineFunction),
    MapConstructor(MapConstructor),
    ArrayConstructor(ArrayConstructor),
    UnaryLookup(UnaryLookup),
}

#[derive(Debug, Clone, PartialEq, Eq)]
#[cfg_attr(test, derive(serde::Serialize))]
pub enum UnaryLookup {
    Name(String),
    IntegerLiteral(i64),
    Expr(ExprS),
}

#[derive(Debug, Clone, PartialEq, Eq)]
#[cfg_attr(test, derive(serde::Serialize))]
pub struct BinaryExpr {
    pub operator: BinaryOperator,
    pub left: PathExpr,
    pub right: PathExpr,
}

#[derive(Debug, Clone, PartialEq, Eq)]
#[cfg_attr(test, derive(serde::Serialize))]
pub struct ApplyExpr {
    pub path_expr: PathExpr,
    pub operator: ApplyOperator,
}

#[derive(Debug, Clone, PartialEq, Eq)]
#[cfg_attr(test, derive(serde::Serialize))]
pub enum ApplyOperator {
    SimpleMap(Vec<PathExpr>),
    Unary(Vec<UnaryOperator>),
    Arrow(Vec<(ArrowFunctionSpecifier, Vec<ExprSingleS>)>),
    Cast(SingleType),
    Castable(SingleType),
    Treat(SequenceType),
    InstanceOf(SequenceType),
}

#[derive(Debug, Clone, PartialEq, Eq)]
#[cfg_attr(test, derive(serde::Serialize))]
pub enum UnaryOperator {
    Plus,
    Minus,
}

#[derive(Debug, Clone, PartialEq, Eq)]
#[cfg_attr(test, derive(serde::Serialize))]
pub struct SingleType {
    pub name: NameS,
    pub question_mark: bool,
}

#[derive(Debug, Clone, PartialEq, Eq)]
#[cfg_attr(test, derive(serde::Serialize))]
pub enum ArrowFunctionSpecifier {
    Name(EQName),
    VarRef(EQName),
    Expr(ExprS),
}

#[derive(Debug, Clone, PartialEq, Eq)]
#[cfg_attr(test, derive(serde::Serialize))]
pub struct MapConstructor {
    pub entries: Vec<MapConstructorEntry>,
}

#[derive(Debug, Clone, PartialEq, Eq)]
#[cfg_attr(test, derive(serde::Serialize))]
pub struct MapConstructorEntry {
    pub key: ExprSingleS,
    pub value: ExprSingleS,
}

#[derive(Debug, Clone, PartialEq, Eq)]
#[cfg_attr(test, derive(serde::Serialize))]
pub struct ArrayConstructor {
    pub members: Vec<ExprSingleS>,
}

#[derive(Debug, Clone, PartialEq, Eq)]
#[cfg_attr(test, derive(serde::Serialize))]
pub enum Literal {
    Decimal(Decimal),
    Integer(IBig),
    Double(OrderedFloat<f64>),
    String(String),
}

#[derive(Debug, Clone, PartialEq, Eq)]
#[cfg_attr(test, derive(serde::Serialize))]
pub struct FunctionCall {
    pub name: Name,
    pub arguments: Vec<ExprSingleS>,
}

#[derive(Debug, Clone, PartialEq, Eq)]
#[cfg_attr(test, derive(serde::Serialize))]
pub struct NamedFunctionRef {
    pub name: NameS,
    pub arity: u8,
}

#[derive(Debug, Clone, PartialEq, Eq)]
#[cfg_attr(test, derive(serde::Serialize))]
pub struct InlineFunction {
    pub params: Vec<Param>,
    pub return_type: Option<SequenceType>,
    pub body: Option<ExprS>,
}

// a function signature as described by:
// https://www.w3.org/TR/xpath-functions-31/#func-signatures
#[derive(Debug, Clone, PartialEq, Eq)]
#[cfg_attr(test, derive(serde::Serialize))]
pub struct Signature {
    pub name: Name,
    pub params: Vec<SignatureParam>,
    pub return_type: SequenceType,
}

#[derive(Debug, Clone, PartialEq, Eq)]
#[cfg_attr(test, derive(serde::Serialize))]
pub struct Param {
    pub name: Name,
    pub type_: Option<SequenceType>,
}

#[derive(Debug, Clone, PartialEq, Eq)]
#[cfg_attr(test, derive(serde::Serialize))]
pub struct SignatureParam {
    pub name: Name,
    pub type_: SequenceType,
}

#[derive(Debug, Clone, PartialEq, Eq)]
#[cfg_attr(test, derive(serde::Serialize))]
pub enum Postfix {
    // vec contains at least 1 element
    Predicate(ExprS),
    ArgumentList(Vec<ExprSingleS>),
    Lookup(Lookup),
}

#[derive(Debug, Clone, PartialEq, Eq)]
#[cfg_attr(test, derive(serde::Serialize))]
pub enum Lookup {
    Name(String),
    IntegerLiteral(i64),
    Expr(Vec<ExprSingleS>),
    Star,
}

#[derive(Debug, Clone, PartialEq, Eq)]
#[cfg_attr(test, derive(serde::Serialize))]
pub struct PathExpr {
    pub steps: Vec<StepExprS>,
}

#[derive(Debug, Clone, PartialEq, Eq)]
#[cfg_attr(test, derive(serde::Serialize))]
pub enum StepExpr {
    PrimaryExpr(PrimaryExprS),
    PostfixExpr {
        primary: PrimaryExprS,
        postfixes: Vec<Postfix>,
    },
    AxisStep(AxisStep),
}

#[derive(Debug, Clone, PartialEq, Eq)]
#[cfg_attr(test, derive(serde::Serialize))]
pub struct AxisStep {
    pub axis: Axis,
    pub node_test: NodeTest,
    pub predicates: Vec<ExprS>,
}

#[derive(Debug, Clone, PartialEq, Eq)]
#[cfg_attr(test, derive(serde::Serialize))]
pub enum Axis {
    Ancestor,
    AncestorOrSelf,
    Attribute,
    Child,
    Descendant,
    DescendantOrSelf,
    Following,
    FollowingSibling,
    Namespace,
    Parent,
    Preceding,
    PrecedingSibling,
    Self_,
}

#[derive(Debug, Clone, PartialEq, Eq)]
#[cfg_attr(test, derive(serde::Serialize))]
pub enum NodeTest {
    KindTest(KindTest),
    NameTest(NameTest),
}

#[derive(Debug, Clone, PartialEq, Eq)]
#[cfg_attr(test, derive(serde::Serialize))]
pub enum NameTest {
    Name(Name),
    Star,
    LocalName(String),
    Namespace(String),
}

#[derive(Debug, Clone, PartialEq, Eq)]
#[cfg_attr(test, derive(serde::Serialize))]
pub enum EQName {
    QName(QName),
    URIQualifiedName(URIQualifiedName),
}

#[derive(Debug, Clone, PartialEq, Eq)]
#[cfg_attr(test, derive(serde::Serialize))]
pub enum QName {
    PrefixedName(PrefixedName),
    UnprefixedName(UnprefixedName),
}

#[derive(Debug, Clone, PartialEq, Eq)]
#[cfg_attr(test, derive(serde::Serialize))]
pub struct PrefixedName {
    pub prefix: String,
    pub local_part: String,
}

#[derive(Debug, Clone, PartialEq, Eq)]
#[cfg_attr(test, derive(serde::Serialize))]
pub struct UnprefixedName {
    pub local_part: String,
}

#[derive(Debug, Clone, PartialEq, Eq)]
#[cfg_attr(test, derive(serde::Serialize))]
pub struct URIQualifiedName {
    pub uri: String,
    pub local_part: String,
}

#[derive(Debug, Clone, PartialEq, Eq)]
#[cfg_attr(test, derive(serde::Serialize))]
pub enum SequenceType {
    Empty,
    Item(Item),
    Unsupported,
}

#[derive(Debug, Clone, PartialEq, Eq)]
#[cfg_attr(test, derive(serde::Serialize))]
pub struct Item {
    pub item_type: ItemType,
    pub occurrence: Occurrence,
}

#[derive(Debug, Clone, PartialEq, Eq)]
#[cfg_attr(test, derive(serde::Serialize))]
pub enum ItemType {
    Item,
    AtomicOrUnionType(NameS),
    KindTest(KindTest),
    FunctionTest(Box<FunctionTest>),
    MapTest(Box<MapTest>),
    ArrayTest(Box<ArrayTest>),
}

impl ItemType {
    pub fn is_generalized_atomic_type(&self) -> bool {
        matches!(self, ItemType::AtomicOrUnionType(_))
    }
}

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
#[cfg_attr(test, derive(serde::Serialize))]
pub enum Occurrence {
    One,
    Option,
    Many,
    NonEmpty,
}

#[derive(Debug, Clone, PartialEq, Eq)]
#[cfg_attr(test, derive(serde::Serialize))]
pub enum KindTest {
    Document(Option<DocumentTest>),
    Element(Option<ElementTest>),
    Attribute(Option<AttributeTest>),
    SchemaElement(SchemaElementTest),
    SchemaAttribute(SchemaAttributeTest),
    PI(Option<PITest>),
    Comment,
    Text,
    NamespaceNode,
    Any,
}

#[derive(Debug, Clone, PartialEq, Eq)]
#[cfg_attr(test, derive(serde::Serialize))]
pub enum DocumentTest {
    Element(ElementTest),
    SchemaElement(SchemaElementTest),
}

#[derive(Debug, Clone, PartialEq, Eq)]
#[cfg_attr(test, derive(serde::Serialize))]
pub struct ElementTest {
    pub name_test: ElementNameOrWildcard,
    pub type_name: Option<ElementTypeName>,
}

#[derive(Debug, Clone, PartialEq, Eq)]
#[cfg_attr(test, derive(serde::Serialize))]
pub struct ElementTypeName {
    pub name: Name,
    pub question_mark: bool,
}

#[derive(Debug, Clone, PartialEq, Eq)]
#[cfg_attr(test, derive(serde::Serialize))]
pub enum ElementNameOrWildcard {
    Name(Name),
    Wildcard,
}

#[derive(Debug, Clone, PartialEq, Eq)]
#[cfg_attr(test, derive(serde::Serialize))]
pub struct AttributeTest {
    pub name_test: AttribNameOrWildcard,
    pub type_name: Option<Name>,
}

#[derive(Debug, Clone, PartialEq, Eq)]
#[cfg_attr(test, derive(serde::Serialize))]
pub enum AttribNameOrWildcard {
    Name(Name),
    Wildcard,
}

#[derive(Debug, Clone, PartialEq, Eq)]
#[cfg_attr(test, derive(serde::Serialize))]
pub struct SchemaElementTest {
    pub name: Name,
}

#[derive(Debug, Clone, PartialEq, Eq)]
#[cfg_attr(test, derive(serde::Serialize))]
pub struct SchemaAttributeTest {
    pub name: Name,
}

#[derive(Debug, Clone, PartialEq, Eq)]
#[cfg_attr(test, derive(serde::Serialize))]
pub enum FunctionTest {
    AnyFunctionTest,
    TypedFunctionTest(TypedFunctionTest),
}

#[derive(Debug, Clone, PartialEq, Eq)]
#[cfg_attr(test, derive(serde::Serialize))]
pub struct TypedFunctionTest {
    parameter_types: Vec<SequenceType>,
    return_type: SequenceType,
}

#[derive(Debug, Clone, PartialEq, Eq)]
#[cfg_attr(test, derive(serde::Serialize))]
pub enum MapTest {
    AnyMapTest,
    TypedMapTest(TypedMapTest),
}

#[derive(Debug, Clone, PartialEq, Eq)]
#[cfg_attr(test, derive(serde::Serialize))]
pub struct TypedMapTest {
    pub key_type: Name,
    pub value_type: SequenceType,
}

#[derive(Debug, Clone, PartialEq, Eq)]
#[cfg_attr(test, derive(serde::Serialize))]
pub enum ArrayTest {
    AnyArrayTest,
    TypedArrayTest(TypedArrayTest),
}

#[derive(Debug, Clone, PartialEq, Eq)]
#[cfg_attr(test, derive(serde::Serialize))]
pub struct TypedArrayTest {
    pub item_type: SequenceType,
}

#[derive(Debug, Clone, PartialEq, Eq)]
#[cfg_attr(test, derive(serde::Serialize))]
pub enum PITest {
    Name(String),
    StringLiteral(String),
}
