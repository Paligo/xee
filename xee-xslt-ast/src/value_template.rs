use crate::ast_core::Span;

struct ValueTemplateTokenizer<'a> {
    s: &'a str,
    char_indices: std::iter::Peekable<std::str::CharIndices<'a>>,
    span: Span,
    mode: Mode,
    start: usize,
    done: bool,
}

enum Mode {
    String,
    Value,
    StartCurly,
    EndCurly,
}

impl<'a> ValueTemplateTokenizer<'a> {
    fn new(s: &'a str, span: Span) -> Self {
        Self {
            s,
            char_indices: s.char_indices().peekable(),
            span,
            mode: Mode::String,
            start: 0,
            done: false,
        }
    }

    fn span(&self, start: usize, end: usize) -> Span {
        Span {
            start: self.span.start + start,
            end: self.span.start + end,
        }
    }

    fn string_item(
        &mut self,
        start: usize,
        end: usize,
    ) -> Option<Result<ValueTemplateItem<'a>, Error>> {
        if let Some(text) = self.s.get(start..end) {
            if text.is_empty() {
                return self.next();
            }
            Some(Ok(ValueTemplateItem::String {
                text,
                span: self.span(start, end),
            }))
        } else {
            Some(Err(Error::IllegalSlice))
        }
    }

    fn value_item(&self, start: usize, end: usize) -> Result<ValueTemplateItem<'a>, Error> {
        if let Some(text) = self.s.get(start..end) {
            Ok(ValueTemplateItem::Value {
                text,
                span: self.span(start, end),
            })
        } else {
            Err(Error::IllegalSlice)
        }
    }

    fn start_curly_item(&self) -> Result<ValueTemplateItem<'a>, Error> {
        Ok(ValueTemplateItem::Curly { c: '{' })
    }

    fn end_curly_item(&self) -> Result<ValueTemplateItem<'a>, Error> {
        Ok(ValueTemplateItem::Curly { c: '}' })
    }
}

#[derive(Debug, PartialEq, Eq)]
#[cfg_attr(feature = "serde", derive(serde::Serialize))]
enum ValueTemplateItem<'a> {
    String { text: &'a str, span: Span },
    Curly { c: char },
    Value { text: &'a str, span: Span },
}

#[derive(Debug, PartialEq, Eq)]
#[cfg_attr(feature = "serde", derive(serde::Serialize))]
enum Error {
    UnescapedCurly { c: char, span: Span },
    IllegalSlice,
}

impl<'a> Iterator for ValueTemplateTokenizer<'a> {
    type Item = Result<ValueTemplateItem<'a>, Error>;

    fn next(&mut self) -> Option<Result<ValueTemplateItem<'a>, Error>> {
        if self.done {
            return None;
        }
        let start = self.start;
        let end;
        match self.mode {
            Mode::String => loop {
                if let Some((i, c)) = self.char_indices.next() {
                    match c {
                        '{' => {
                            self.mode = Mode::StartCurly;
                            end = i;
                            self.start = end + 1;
                            return self.string_item(start, end);
                        }
                        '}' => {
                            self.mode = Mode::EndCurly;
                            end = i;
                            self.start = end + 1;
                            return self.string_item(start, end);
                        }
                        _ => {
                            continue;
                        }
                    }
                } else {
                    end = self.s.len();
                    self.done = true;
                    return self.string_item(start, end);
                }
            },
            Mode::Value => loop {
                // TODO: this should parse an XPath expression
                // with the following string and return the index
                // of the end of the expression, which should contain
                // a close curly brace or be an error
                // this should be fed into the lexer in an iterative way

                if let Some((i, c)) = self.char_indices.next() {
                    if c == '}' {
                        self.mode = Mode::String;
                        end = i;
                        self.start = end + 1;
                        return Some(self.value_item(start, end));
                    }
                } else {
                    self.done = true;
                    return Some(Err(Error::UnescapedCurly {
                        c: '{',
                        span: Span {
                            start: self.span.start + self.start,
                            end: self.span.start + self.start + 1,
                        },
                    }));
                }
            },
            Mode::StartCurly => {
                if let Some((_peek_i, peek_c)) = self.char_indices.peek() {
                    if *peek_c == '{' {
                        if let Some((i, _c)) = self.char_indices.next() {
                            self.mode = Mode::String;
                            end = i + 1;
                            self.start = end;
                            return Some(self.start_curly_item());
                        } else {
                            unreachable!();
                        }
                    }
                }
                self.mode = Mode::Value;
                self.next()
            }
            Mode::EndCurly => {
                if let Some((_peek_i, peek_c)) = self.char_indices.peek() {
                    if *peek_c == '}' {
                        if let Some((i, _c)) = self.char_indices.next() {
                            self.mode = Mode::String;
                            end = i + 1;
                            self.start = end;
                            return Some(self.end_curly_item());
                        } else {
                            unreachable!();
                        }
                    }
                }
                self.done = true;
                Some(Err(Error::UnescapedCurly {
                    c: '}',
                    span: Span {
                        start: self.span.start + self.start,
                        end: self.span.start + self.start + 1,
                    },
                }))
            }
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use insta::assert_ron_snapshot;

    fn parse_with_span(s: &str, span: Span) -> Result<Vec<ValueTemplateItem>, Error> {
        let tokenizer = ValueTemplateTokenizer::new(s, span);
        tokenizer.collect()
    }

    fn parse(s: &str) -> Result<Vec<ValueTemplateItem>, Error> {
        let span = Span {
            start: 0,
            end: s.len(),
        };
        parse_with_span(s, span)
    }

    #[test]
    fn test_string_without_curly() {
        assert_ron_snapshot!(parse("hello world"));
    }

    #[test]
    fn test_string_start_curly_escaped() {
        assert_ron_snapshot!(parse("hello{{world"));
    }

    #[test]
    fn test_string_end_curly_escaped() {
        assert_ron_snapshot!(parse("hello}}world"));
    }

    #[test]
    fn test_string_with_value() {
        assert_ron_snapshot!(parse("hello {world}!"));
    }

    #[test]
    fn test_string_with_value_in_span() {
        let s = "hello {world}!";
        let span = Span {
            start: 10,
            end: s.len() + 10,
        };
        assert_ron_snapshot!(parse_with_span(s, span));
    }

    #[test]
    fn test_string_with_empty_value() {
        assert_ron_snapshot!(parse("hello {}!"));
    }

    #[test]
    fn test_string_with_multiple_values() {
        assert_ron_snapshot!(parse("hello {a} and {b}!"));
    }

    #[test]
    fn test_string_with_multiple_adjacent_values() {
        assert_ron_snapshot!(parse("hello {a}{b}!"));
    }

    #[test]
    fn test_string_unescaped_unclosed_start_curly() {
        assert_ron_snapshot!(parse("hello{world"));
    }

    #[test]
    fn test_string_unescaped_unclosed_start_curly_with_span() {
        let s = "hello{world";
        let span = Span {
            start: 10,
            end: 10 + s.len(),
        };
        assert_ron_snapshot!(parse_with_span(s, span));
    }

    #[test]
    fn test_string_unescaped_end_curly() {
        assert_ron_snapshot!(parse("hello}world"));
    }
}
