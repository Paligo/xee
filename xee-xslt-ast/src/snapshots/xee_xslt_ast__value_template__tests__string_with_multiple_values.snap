---
source: xee-xslt-ast/src/value_template.rs
expression: "parse(\"hello {a} and {b}!\")"
---
Ok([
  String(
    text: "hello ",
    span: Span(
      start: 0,
      end: 6,
    ),
  ),
  Value(
    text: "a",
    span: Span(
      start: 7,
      end: 8,
    ),
  ),
  String(
    text: " and ",
    span: Span(
      start: 9,
      end: 14,
    ),
  ),
  Value(
    text: "b",
    span: Span(
      start: 15,
      end: 16,
    ),
  ),
  String(
    text: "!",
    span: Span(
      start: 17,
      end: 18,
    ),
  ),
])
